### ISSUE TITLE
webext-test-functional is broken with Node 16

### ISSUE THREAD

**Issue by willdurand** [mozilla/addons-linter#3788]
See: https://app.circleci.com/pipelines/github/mozilla/addons-linter/1123/workflows/88bf9a41-2c8d-4cc8-a6fc-c2db9ced1d03/jobs/2538

Output from Circle CI:

```
  web-ext build
    ✔ should accept: --source-dir SRCDIR (240ms)
    ✔ throws an error on multiple -n (163ms)

  web-ext lint
This test failed. Please check the log below to debug.
    1) should accept: --source-dir SRCDIR

  web-ext
    ✔ should accept: --help (200ms)
    ✔ should hide --input from --help output (190ms)

  web-ext run
    ✔ accepts: --no-reload --watch-file --watch-files --source-dir SRCDIR --firefox FXPATH --watch-ignored (539ms)
    ✔ should not accept: --watch-file <directory> (605ms)
    ✔ should not accept: --target INVALIDTARGET (185ms)

  web-ext sign
    ✔ should accept: --source-dir SRCDIR --api-url-prefix URL (699ms)
    ✔ should use config file if required parameters are not in the arguments (478ms)
    ✔ should show an error message if the api-key is not set in the config (303ms)
    ✔ should show an error message if the api-key cannot be found (219ms)

  web-ext imported as a library
    ✔ can be imported as an ESM module (2182ms)
    ✔ can be imported as a CommonJS module (2136ms)

  web-ext
    ✔ recommends matching command (159ms)


  14 passing (12s)
  1 failing

  1) web-ext lint
       should accept: --source-dir SRCDIR:
     Error: Unexpected web-ext functional test result:
argv:
  - lint
  - --source-dir
  - /tmp/tmp-web-ext--487-EqWXkpg4kOb7/tmp-addon-dir
  - --verbose
exitCode: 1
stdout:
  """
    [program.js][info] Version: master-1e265f85966268b650c6d6be94842c8a0cf45f74
    [program.js][debug] Discovering config files. Set --no-config-discovery to disable
    [config.js][debug] Discovered config "/home/circleci/.web-ext-config.js" does not exist or is not readable
    [config.js][debug] Discovered config "/tmp/tmp-web-ext--487-EqWXkpg4kOb7/package.json" does not exist or is not readable
    [config.js][debug] Discovered config "/tmp/tmp-web-ext--487-EqWXkpg4kOb7/web-ext-config.js" does not exist or is not readable
    [program.js][error]
    Error: Cannot find module 'yargs'
    Require stack:
    - /home/circleci/addons-linter/dist/addons-linter.js
    - /home/circleci/addons-linter/web-ext/dist/web-ext.js
    - /home/circleci/addons-linter/web-ext/bin/web-ext
        at Function.Module._resolveFilename (node:internal/modules/cjs/loader:941:15)
        at Function.Module._load (node:internal/modules/cjs/loader:774:27)
        at Module.require (node:internal/modules/cjs/loader:1013:19)
        at require (node:internal/modules/cjs/helpers:93:18)
        at Array.Object.defineProperty.value (/home/circleci/addons-linter/dist/webpack:/addons-linter/external "yargs":1:1)
        at __webpack_require__ (/home/circleci/addons-linter/dist/webpack:/addons-linter/webpack/bootstrap:19:1)
        at Array.<anonymous> (/home/circleci/addons-linter/dist/webpack:/addons-linter/src/cli.js:1:1)
        at __webpack_require__ (/home/circleci/addons-linter/dist/webpack:/addons-linter/webpack/bootstrap:19:1)
        at Array.Object.defineProperty.value (/home/circleci/addons-linter/dist/webpack:/addons-linter/src/main.js:1:1)
        at __webpack_require__ (/home/circleci/addons-linter/dist/webpack:/addons-linter/webpack/bootstrap:19:1)

    [program.js][error] Error code: MODULE_NOT_FOUND

    [program.js][debug] Command executed: lint

  """
stderr:
      at reportCommandErrors (tests/functional/common.js:73:17)
      at /home/circleci/addons-linter/web-ext/tests/functional/test.cli.lint.js:16:12
      at processTicksAndRejections (node:internal/process/task_queues:96:5)




Exited with code exit status 1
```

It's not clear what's going on right now but we should understand the problem and fix it.

**Comment by rpl** [comment]
@willdurand I was able to reproduce this issue locally and so I took a more deeper look into it.
 
I reproduced it inside a docker container using the `circleci` CLI utility, to ensure the test was going to run in an environment as similar as possible as the one where it does fail in the CI job.

The main different with what circleci does is the cached npm directory, which is not supported when running the circleci config locally on docker, and so we can exclude that from the possible causes of the issue (also the script that pull the web-ext sources and run web-ext integration tests does run npm install again and in my opinion that also exclude the circleci cache from the possible causes).

After some digging, I'm pretty sure that the issue is the way our `webext-test-functional` script is trying to simulate running web-ext functional tests on the addons-linter sources through `npm link`.

My guess is that when the script is calling `npm link addons-linter` we end up with a node_modules directory that is triggering the "yargs module resolution" issue.

To double-check if that was the case, I tried locally to tweak the `webext-test-functional` script to use [`fyn add`](https://www.npmjs.com/package/fyn#fynlocal-mode) instead of `npm link` and with those few changes the circleci job was passing just fine.

The following are the changes applied:
```diff
diff --git a/scripts/webext-test-functional b/scripts/webext-test-functional
index 83d5e57c..5a84094b 100755
--- a/scripts/webext-test-functional
+++ b/scripts/webext-test-functional
@@ -4,17 +4,19 @@
 npm install
 npm run build
 
-# Create (global) addons-linter link
-# See: https://docs.npmjs.com/cli/link
-npm link
+LINTER_PATH=`pwd`
 
 # Fetch web-ext
 git clone https://github.com/mozilla/web-ext
 cd web-ext
 
-# Link addons-linter, then install web-ext's dependencies
+# install web-ext's dependencies, then link the addons-linter
+# dependency from the local addons-linter repo clone.
+# NOTE: this is using `fyn add` instead of `npm link` to ensure
+# that web-ext will resolve addons-linter dependencies similarly
+# to when it is installed from the npm registry (see #3788).
 npm install
-npm link addons-linter
+npx fyn add $LINTER_PATH
 
 # Run web-ext functional tests
 MOCHA_TIMEOUT=60000 npm run test-functional
```

I propose the following next steps:

- confirm that the changes in the diff would make the job to pass on circleci too as they did for me locally, we can just try to add this changes in #3787 temporarily in a separate patch added in that branch (and uncomment the line that should run the webext functional tests, which is currently commended out because of this issue)

- If that is the case (it should be, but better be sure of it) then we may evaluate if we would be ok with using `npx fyn add` for this or if we want to dig more
 
To be honest I'm not sure it is really worth, mostly because I think that `npm link` is making our test environment a bit "artificial" and not really matching what a user would experience while installing web-ext from npm, nevertheless I'm definitely open to dig more if we agreed that it is what we prefer.

wdyt?

**Comment by rpl** [comment]
On a second thought, given that we run this test on in the CI and it is a bash script, there is another workaround that would work (I also double-checked that it does work as I was assuming by running it locally inside the circleci docker images) and would not require downloading other npm package (I mean the `npx fyn` from the workaround in my previous comment):

We could just create (inside the web-ext node_modules dir) a symbolic link to the addons-linter repo working tree (which is initialized explicitly with npm install and so the addons-linter dependencies should be resolved just fine).

Here is the diff for this alternative workaround:
```diff
diff --git a/scripts/webext-test-functional b/scripts/webext-test-functional
index 83d5e57c..763808c9 100755
--- a/scripts/webext-test-functional
+++ b/scripts/webext-test-functional
@@ -4,17 +4,17 @@
 npm install
 npm run build
 
-# Create (global) addons-linter link
-# See: https://docs.npmjs.com/cli/link
-npm link
+LINTER_PATH=`pwd`
 
 # Fetch web-ext
 git clone https://github.com/mozilla/web-ext
 cd web-ext
 
-# Link addons-linter, then install web-ext's dependencies
+# install web-ext's dependencies, then link the addons-linter
+# dependency from the local addons-linter repo clone.
 npm install
-npm link addons-linter
+mv node_modules/addons-linter node_modules/addons-linter.npm
+ln -s $LINTER_PATH node_modules/addons-linter
 
 # Run web-ext functional tests
 MOCHA_TIMEOUT=60000 npm run test-functional
```

Personally I think that this would be a reasonable workaround even in the long run.

Let me know wdyt.

**Comment by willdurand** [comment]
> Personally I think that this would be a reasonable workaround even in the long run.

Yes, same but isn't it what `npm link` is supposed to do? 🤔

**Comment by rpl** [comment]
> > Personally I think that this would be a reasonable workaround even in the long run.
> 
> Yes, same but isn't it what `npm link` is supposed to do? thinking

well, given that it does lead to different results, I would guess that in practice `npm link` does something different (at least in nodejs 16 where it does fail, if I'm not mistaken the test was not failing on the other nodejs versions, right?) and that seems to affects the dependencies resolutions.

If we think it would be worth to dig into it more, I think this source file would be the starting point to look more deeply into what `npm link` does differently: https://github.com/npm/cli/blob/latest/lib/link.js

**Comment by willdurand** [comment]
> If we think it would be worth to dig into it more, I think this source file would be the starting point to look more deeply into what `npm link` does differently: [https://github.com/npm/cli/blob/latest/lib/link.js](https://github.com/npm/cli/blob/latest/lib/link.js?rgh-link-date=2021-06-24T11%3A03%3A32Z)

ugh, r+ on your patch above 😅

**Comment by willdurand** [comment]
Fixed by https://github.com/mozilla/addons-linter/pull/3787
